[YUNIKORN-2012] group quota is not considered during pod scheduling

manirajv06 opened the pull request:
### What is this PR for?

Problem: When config has limits set only for group, its quota was not getting honoured. User belonging to the group are allowed to use more resources than the allowed limit.

Besides fixing the issue, this pr also simplifies the "cleaning up old user/group limits resource usage when config changes" codepath by removing the unnecessary clean up processing for the queue path. For example, say group "test" limit is set for queue path a.b.c. A config change had happened for the removal of "test" from a.b.c. So, clean up is required only when a.b.c queue path comes. But, clean up happens even when we are processing a & b from queue configs just because those queues has been tracked for the group "test". So far, as long as the queue has been tracked for a group, then, even that queue enters the clean up loop. This unnecessary processing reset the values already set not only for the a & b, even for c. which is not needed and correct.

Now simplified the whole codepath by comparing the old config with the new config. If new config doesn't have the limit set for group "test", then clean up triggers.


### What type of PR is it?
* [ ] - Bug Fix

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-2012

### How should this be tested?

### Screenshots (if appropriate)

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

manirajv06 commented:
Notes:

Instead of locking objects for longer duration, moved the locks to the nearest place so that scheduler performance doesn't degrade during config changes. Changed the whole approach of clearing old configs by storing the new configs in separate maps and do the clearance only comparing the existing configs with new configs. For the newer limit configs, things happen while during the traversal and storing in new configs itself. Once all the work is over, replace the existing configs with new configs by keeping the write lock. This way, mostly read lock happens on the existing configs and write lock only during the final switch over.

In addition, some more hidden issues has been identified and fixed them too.

1. When group limits removed in new config (root.parent.leaf) and contains the same group config in another level (root.parent),  during the clearance step, we reduce/reset the resource usage for the entire group queue hierarchy (root.parent.leaf). i.e for root.parent though it has same group limit configs.
2. Changes in unlinking queue trackers.

wilfred-s commented on pkg/scheduler/ugm/manager.go:
```diff
@@ -365,127 +381,159 @@ func (m *Manager) internalProcessConfig(cur configs.QueueConfig, queuePath strin
+	if len(m.userLimits) == 0 {
+		return
+	}
+	m.RLock()
+	defer m.RUnlock()
```
needs to be before the check of the userlimits

wilfred-s commented on pkg/scheduler/ugm/manager.go:
```diff
@@ -365,127 +381,159 @@ func (m *Manager) internalProcessConfig(cur configs.QueueConfig, queuePath strin
-			for app, user := range appUsersMap {
-				ut := m.userTrackers[user]
-				ut.setGroupForApp(app, nil)
+func (m *Manager) clearEarlierSetUserLimits(newUserLimits map[string]map[string]*LimitConfig) {
```
keep the comment

wilfred-s commented on pkg/scheduler/ugm/manager.go:
```diff
@@ -365,127 +381,159 @@ func (m *Manager) internalProcessConfig(cur configs.QueueConfig, queuePath strin
 }
 
-func (m *Manager) clearEarlierSetUserLimits(ut *UserTracker, queuePath string, userLimits map[string]bool) {
+func (m *Manager) clearEarlierSetUserLimit(ut *UserTracker, queuePath string) {
```
This really removes, while the previous function does the looping:
the names look too much alike difficult to understand the difference between the two.

wilfred-s commented on pkg/scheduler/ugm/manager.go:
```diff
@@ -365,127 +381,159 @@ func (m *Manager) internalProcessConfig(cur configs.QueueConfig, queuePath strin
+	}
+}
+
+func (m *Manager) clearEarlierSetGroupLimits(newGroupLimits map[string]map[string]*LimitConfig) {
```
Similar comment as for the user: comments for the functions to explain what they do and the names look too much alike

wilfred-s commented on pkg/scheduler/ugm/manager.go:
```diff
@@ -365,127 +381,159 @@ func (m *Manager) internalProcessConfig(cur configs.QueueConfig, queuePath strin
+	if len(m.groupLimits) == 0 {
+		return
+	}
+	m.RLock()
+	defer m.RUnlock()
```
move before the check

manirajv06 commented:
@wilfred-s Addressed review comments. Also, tried running #669 pr test changes with this pr and can see tests passing through without any issues. Can you take a final look and approve?

wilfred-s reviewed (approved):
LGTM
